pysindy/optimizers/sbr.py

```diff
@@ -0,0 +1,163 @@
+import jax.numpy as jnp
+import numpy as np
+import numpyro
+from jax import random
+from numpyro.distributions import Exponential
+from numpyro.distributions import HalfCauchy
+from numpyro.distributions import InverseGamma
+from numpyro.distributions import Normal
+from numpyro.infer import MCMC
+from numpyro.infer import NUTS
+
+from .base import BaseOptimizer
+
+
+class SBR(BaseOptimizer):
+    """
+    Sparse Bayesian Regression (SBR) optimizer. This uses the regularised
+    horseshoe prior over the SINDy coefficients to achieve sparsification.
+
+    Inference is achieved through NUTS sampling with numpyro, which is a
+    dependency for this optimizer.
+
+    The SINDy coefficients are set as the posterior means of the MCMC samples.
+    Additional statistics can be computed from the MCMC samples stored in
+    the mcmc attribute using e.g. ArviZ.
+
+    See the following reference for more details:
+
+        Hirsh, S. M., Barajas-Solano, D. A., & Kutz, J. N. (2021).
+        parsifying Priors for Bayesian Uncertainty Quantification in
+        Model Discovery (arXiv:2107.02107). arXiv. http://arxiv.org/abs/2107.02107
+
+    Parameters
+    ----------
+    sparsity_coef_tau0 : float, optional (default 0.1)
+        Sparsity coefficient for regularised horseshoe hyper-prior. Lower
+        value increases the sparsity of the SINDy coefficients. See Hirsh
+        et al. (2021) for more details.
+
+    slab_shape_nu : float, optional (default 4)
+        Shape parameter for the slab region of the regularised horseshoe
+        prior. See Hirsh et al. (2021) for more details
+
+    slab_shape_s : float, optional (default 2)
+        Shape parameter for the slab region of the regularised horseshoe
+        prior. See Hirsh et al. (2021) for more details
+
+    noise_hyper_lambda : float, optional (default 1)
+        Rate hyperparameter for the exponential prior distribution over
+        the noise standard deviation.
```

Comment:
@Jacob-Stevens-Haas I'm happy with this. The `noise_hyper_lambda` is the hyperparameter for the exponential distribution over the observation noise parameter `sigma`. So it's not relevant to the half-cauchys for the parameters, only the Gaussian likelihood for the data.

Reply:
Thanks!  OK I see equation 3.3/3.4 "We must choose priors for the noise level parameters $σ_j$".  

Soooo this gets dramatically more exciting, but the tl;dr: is that the docstring for it is great, maybe just name `noise_hyper_lambda` something like `noise_hyper`, since the paper didn't use a greek letter here.  I'll have some more questions about the MCMC implementation, but maybe dial the default way down (is zero legal?), since (a) the case of recovering an exact simulation without noise is perhaps the most common, and (b) if i read it correctly (check me here) the paper uses different hypers & distributions for measurement variance prior: $\sigma_u, \sigma_\nu \sim LogNormal(\mu=-1, \sigma=.1)$ in § 3.2.1, $\sigma_u, \sigma_\nu \sim Gamma(\alpha=1, \beta=1)$ in § 3.2.2, and unspecified in § 3.2.3.  What are your thoughts?